File: ios/NativeSlotStore.swift
import Foundation

final class NativeSlotStore {
    private(set) var enabled = false
    private var slots: [String: NativeSlotState] = [:]

    func setEnabled(_ value: Bool) {
        enabled = value
        if !enabled {
            slots.removeAll()
        }
    }

    func put(placementId: String, slotId: String, hostId: String, adUnitId: String, ttlMs: Double) {
        slots[slotId] = NativeSlotState(
            placementId: placementId,
            slotId: slotId,
            hostId: hostId,
            adUnitId: adUnitId,
            ttlMs: ttlMs,
            createdAtMs: Date().timeIntervalSince1970 * 1000,
            attached: false
        )
    }

    func get(slotId: String) -> NativeSlotState? {
        slots[slotId]
    }

    func update(_ state: NativeSlotState) {
        slots[state.slotId] = state
    }

    func remove(slotId: String) {
        slots.removeValue(forKey: slotId)
    }

    func clear() {
        slots.removeAll()
    }
}

File: android/src/main/java/id/donugr/dgradmobbridge/NativeSlotStore.java
package id.donugr.dgradmobbridge;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

class NativeSlotStore {
    private boolean enabled = false;
    private final Map<String, NativeSlotState> slots = new ConcurrentHashMap<>();

    void setEnabled(boolean value) {
        enabled = value;
        if (!enabled) {
            slots.clear();
        }
    }

    boolean isEnabled() {
        return enabled;
    }

    NativeSlotState getOrCreate(String slotId) {
        return slots.computeIfAbsent(slotId, NativeSlotState::new);
    }

    NativeSlotState put(NativeSlotState state) {
        slots.put(state.slotId, state);
        return state;
    }

    NativeSlotState get(String slotId) {
        return slots.get(slotId);
    }

    boolean contains(String slotId) {
        return slots.containsKey(slotId);
    }

    NativeSlotState remove(String slotId) {
        return slots.remove(slotId);
    }

    void clear() {
        slots.clear();
    }

    Map<String, NativeSlotState> snapshot() {
        return new ConcurrentHashMap<>(slots);
    }
}
